modules/files.nix

```diff
@@ -27,11 +54,23 @@ in
       type = fileType "<envar>HOME</envar>" homeDirectory;
     };
 
+    home-file-defs = mkOption {
```

Comment:
Good catch of this bug! For the "files: collect file definitions in internal list" commit I would suggest the following simpler change instead:
```patch
diff --git a/modules/lib/file-type.nix b/modules/lib/file-type.nix
index d8622d0..6b7263a 100644
--- a/modules/lib/file-type.nix
+++ b/modules/lib/file-type.nix
@@ -109,7 +109,7 @@ in
       };
 
       config = {
-        target = mkDefault name;
+        target = mkDefault (basePath + "/" + name);
         source = mkIf (config.text != null) (
           mkDefault (pkgs.writeTextFile {
             inherit (config) executable text;
diff --git a/modules/misc/xdg.nix b/modules/misc/xdg.nix
index 3e7a3bc..b185e12 100644
--- a/modules/misc/xdg.nix
+++ b/modules/misc/xdg.nix
@@ -90,7 +90,10 @@ in
     })
 
     {
-      home.file = mkMerge [ cfg.configFile cfg.dataFile ];
+      home.file = mkMerge [
+        (attrValues cfg.configFile)
+        (attrValues cfg.dataFile)
+      ];
     }
   ];
 }
```

Reply:
Actually, changing target is a bad idea. It should be sufficient to just do
```patch
diff --git a/modules/misc/xdg.nix b/modules/misc/xdg.nix
index 3e7a3bc..b185e12 100644
--- a/modules/misc/xdg.nix
+++ b/modules/misc/xdg.nix
@@ -90,7 +90,10 @@ in
     })
 
     {
-      home.file = mkMerge [ cfg.configFile cfg.dataFile ];
+      home.file = mkMerge [
+        (attrValues cfg.configFile)
+        (attrValues cfg.dataFile)
+      ];
     }
   ];
 }
```

Reply:
This doesn't work with commit [`disallow setting source and text
simultaneously`](https://github.com/rycee/home-manager/pull/156/commits/e8f9ec633be5b2188405c5051ca5a5d79c7331a2).

In case of `text != null`, the final, evaluated submodule attrsets in, say, `(attrValues cfg.configFile)` simultaneously define a `source` option and have a non null text option. 
When these attrsets are assigned to `home.file`, they are rejected because of the double definition.

Reply:
Ah, that's true, at that point the `source` would have been set. But then it should always be safe to reset the `text` attribute to `null`? I.e., something like

    home.file =
      let
        fixup = files: map (n: n // { text = null; }) (attrValues files);
      in
        fixup cfg.configFile ++ fixup cfg.dataFile;

In any case, at the moment I'm thinking it is best not to remove the possibility of providing both `text` and `source`. My idea is that it removes the possibility for a user to override a `text` that is set in a module. For example, a user for some unimaginable reason wants to enable the bash module but replace the content of the `~/.bash_profile` file that the module writes. The module uses the `text` attribute and the user wants to use `source` like so:

    home.file.".bash_profile".source = ./my.awesome.replacement;

This would cause an error if we disallow simultaneous setting of `text` and `source`. Instead the user would have to do something unpleasant like

    home.file.".bash_profile".text = mkForce (readFile ./my.awesome.replacement);

Admittedly, this example is a bit far-fetched but I can imagine more reasonable examples existing.

Reply:
W.r.t the above I note that a module having

    xdg.configFile."conffile".text = "blah blah";

and a user having

    home.file.".config/conffile".text = "my blah blah"

will not result in the expected override, I believe.